cpmpy/solvers/cplex.py

```diff
@@ -0,0 +1,581 @@
+"""
+import warnings
+from typing import Optional
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..exceptions import NotSupportedError
+from ..expressions.core import *
+from ..expressions.utils import argvals, argval, eval_comparison
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar
+from ..expressions.globalconstraints import DirectConstraint
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.decompose_global import decompose_in_tree
+from ..transformations.flatten_model import flatten_constraint, flatten_objective
+from ..transformations.get_variables import get_variables
+from ..transformations.linearize import linearize_constraint, only_positive_bv, only_positive_bv_wsum
+from ..transformations.normalize import toplevel_list
+from ..transformations.reification import only_implies, reify_rewrite, only_bv_reifies
+from ..transformations.safening import no_partial_functions
+
+class CPM_cplex(SolverInterface):
+    """
+    Interface to the CPLEX solver.
+
+    Creates the following attributes (see parent constructor for more):
+    - cplex_model: object, CPLEX model object
+
+    The :class:`~cpmpy.expressions.globalconstraints.DirectConstraint`, when used, 
+    calls a function on the ``cplex_model`` object.
+
+    Documentation of the solver's own Python API:
+    https://ibmdecisionoptimization.github.io/docplex-doc/mp/docplex.mp.model.html
+    """
+
+    @staticmethod
+    def supported():
+        return CPM_cplex.installed() and CPM_cplex.license_ok()
+
+    @staticmethod
+    def installed():
+        # try to import the package
+        try:
+            import docplex.mp as domp
+        except ModuleNotFoundError as e:
+            warnings.warn(f"CPM_cplex: Could not import docplex: {e}")
+            return False
+        try:
+            import cplex
+            return True
+        except ModuleNotFoundError as e:
+            warnings.warn(f"CPM_cplex: Could not import cplex: {e}")
+            return False
+
+    @staticmethod
+    def license_ok():
+        if not CPM_cplex.installed():
+            warnings.warn(
+                f"License check failed, python package 'docplex' or 'cplex' is not installed! Please check 'CPM_cplex.installed()' before attempting to check license.")
+            return False
+        else:
+            try:
+                from docplex.mp.model import Model
+                mdl = Model()
+                mdl.solve()
+                return True
+            except Exception as e:
+                warnings.warn(f"Problem encountered with CPLEX installation: {e}")
+                return False
+
+    @staticmethod
+    def version() -> Optional[str]:
+        """
+        Returns the installed version of the solver's Python API.
+        """
+        try:
+            import pkg_resources
+            return pkg_resources.get_distribution('docplex').version
```

Comment:
Can you return both the docplex and CPLEX version here?
E.g., such as in CPO:
https://github.com/CPMpy/cpmpy/blob/2bdfcc70bf2a854102bcbdcce30d460f8fb82251/cpmpy/solvers/cpo.py#L123